### DISCUSSION TITLE
Blocking changes from users but allowing on API Route?

### DISCUSSION THREAD

**Discussion by marcusfelix** [pocketbase/pocketbase#256] [Q&A]
Through some tests i have discovered that API rules blocking updates like ```@request.data.status = null``` is effective in API Route as well. How can i accomplish some kind of Admin SDK (Firebase) to override those API rules. Here is what i am trying to do with API routes:

```go
        // Updating project data
	dao := app.Dao()

        // Get project collection
	collection, err := dao.FindCollectionByNameOrId("projects")
	if err != nil {
		return c.JSON(500, map[string]string{"error": "error creating projects model"})
	}
	pid := subscription.Metadata["project"]

	// Get project record
	project, err := dao.FindRecordById(collection, pid, nil)
	if err != nil {
		return c.JSON(500, map[string]string{"error": "error finding project record"})
	}

	// Set updated fields
	project.SetDataValue("status", string(subscription.Status))

	// Update record
	errr := dao.SaveRecord(project)
	if errr != nil {
		return c.JSON(500, map[string]string{"error": "error updating project"})
	}
```

**Comment by ganigeorgiev** [comment]
By default, the collection API rules are ignored if you are authenticated as an admin.

I'm not sure that I understand your example. Are you trying to create a custom API endpoint/route or you want the above code to be executed in a before/after hook? If you create a custom API endpoint/route then its up to the developer to decide whether they want to check the API rules or not.

**Reply by marcusfelix** [reply]
Yes, it is a custom API route. How can i ignore rules updating Records? I think rules are ok on user side, but apparently are blocking inside API route logging ```2022/08/01 13:10:23 sql: no rows in result set```. I'm considering use raw SQL if isn't a better way.

**Reply by ganigeorgiev** [reply]
If it is a custom route then API rules doesn't apply. Make sure that `pid` in the above example actually is defined and a record with the provided id exists.

**Reply by ganigeorgiev** [reply]
@marcusfelix Could you also post your custom route declaration? It is also possible that your action may not be registered properly or you are not accessing it correctly and the router is matching some other route.

**Reply by marcusfelix** [reply]
Sure! 

it is  a Stripe web hook that i manage to get working because GO helps a lot, but i don't know if something is wrong with "project" update step. I'm pretty sure there is a less bloated way to do this, but this code works with more open API rules and fails when a try to block changes on 'status' field made by users with Client SDK.


```go
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
	e.Router.AddRoute(echo.Route{
		Method: http.MethodPost,
		Path:   "/api/events",
		Handler: func(c echo.Context) error {
			const MaxBodyBytes = int64(65536)

			// Getting raw body
			bodyReader := c.Request().Body
			payload, err := ioutil.ReadAll(bodyReader)
			if err != nil {
				return c.JSON(500, map[string]string{"error": "Webhook raw body error"})
			}

			// Checking event
			endpointSecret := os.Getenv("STRIPE_WEBHOOK_SECRET")
			signatureHeader := c.Request().Header.Get("Stripe-Signature")
			event, err := webhook.ConstructEvent(payload, signatureHeader, endpointSecret)
			if err != nil {
				return c.JSON(500, map[string]string{"error": "bad signature"})
			}

			switch event.Type {
			
			case "customer.subscription.updated":
				var subscription stripe.Subscription
				err := json.Unmarshal(event.Data.Raw, &subscription)
				if err != nil {
					return c.JSON(500, map[string]string{"error": "error parsing webhook"})
				}

				// Updating project data
				dao := app.Dao()

				// Get project collection
				collection, err := dao.FindCollectionByNameOrId("projects")
				if err != nil {
					return c.JSON(500, map[string]string{"error": "error creating projects model"})
				}
				pid := subscription.Metadata["project"]

				// Get project record
				project, err := dao.FindRecordById(collection, pid, nil)
				if err != nil {
					return c.JSON(500, map[string]string{"error": "error finding project record"})
				}

				// Get plan name
				plan, err := product.Get(subscription.Plan.Product.ID, nil)
				if err != nil {
					return c.JSON(500, map[string]string{"error": "error getting product details"})
				}

				// Set updated fields
				project.SetDataValue("plan", plan.Name)
				project.SetDataValue("status", string(subscription.Status))

				// Update record
				errr := dao.SaveRecord(project)
				if errr != nil {
					return c.JSON(500, map[string]string{"error": "error updating project"})
				}

	

			default:
				fmt.Fprintf(os.Stderr, "Unhandled event type: %s\n", event.Type)
			}

			return c.JSON(200, map[string]interface{}{})
		},
	})
	return nil
})
```

Thanks!

**Reply by ganigeorgiev** [reply]
> but this code works with more open API rules and fails when a try to block changes on 'status' field from users.

The above action doesn't depend on the collection API rules. You could try sending a POST to `/api/events` and as long as it satisfies your Stripe handling it will be executed.
Furthermore, you are using directly the dao to save your project record which means that even the collection field validations will not be applied (DAO is writing directly to the database; the [RecordUpsert form](https://github.com/pocketbase/pocketbase/blob/master/forms/record_upsert.go) run validation on submit).

So there is definitelly something else going on here, but I have difficulties to understand what do you mean by " fails when a try to block changes on 'status' field from users."

Does it fail before reaching the action? On which error step exactly it fails?
_btw. you could also use the builtin formatted erros from the [rest subpackage](https://github.com/pocketbase/pocketbase/blob/master/tools/rest/api_error.go#L40) (although keep in mind that it will be moved to `apis` in the next release for better discoverability)_

**Reply by marcusfelix** [reply]
I mean that users should not be able to change 'status' directly via REST or SDK. But never mind! I think i mess up with api rules trying to combining set of rules. Now works as i intended. If i understand well, rules acts like filters: If checks falsely, logs 'not found' instead something like 'forbidden'

Thanks for the heads up about error handling and thanks for this amazing project!!

**Reply by ganigeorgiev** [reply]
>  If i understand well, rules acts like filters: If checks falsely, logs 'not found' instead something like 'forbidden'

Yep, that's correct. We cannot return "forbidden" because we don't know on which part of the rule the query will fail. For the list action in case the filters fail an empty `[]` items array is returned, for view/create/update/delete - "404 - not found".

---

> I mean that users should not be able to change 'status' directly via REST or SDK.

If you want to prevent your users to change the value of your `status` field, you could add the following rule to your create and update actions (assuming that the `status` field is marked as required!):
```
// status is REQUIRED field
@request.data.status = '' || status = @request.data.status
```
Which basically translates to:
- `@request.data.status = ''` - allow submitting requests without status (on create the required validator will be triggered, on update if the `status` is not sent with the request it will do nothing)
- `status = @request.data.status` - OR allow requests where the user submitted status matches with the already stored one

Or in other words - the above rule will allow the user to set a `status` value only on create.
If you want to force a specific status on create, just set `status = 'pending'` for example as create API rule.

You also are able to change the status to whatever value you want when authenticated as admin or in a event hook:

```go
app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
    if e.Record.Collection().Name == "projects" {
        if e.Record.GetStringDataValue("status") != "pending" {
            // always set "pending" status on update
            e.Record.SetDataValue("status", "pending")
            // You don't have to save the record here.
            // Any changes to the e.Record will be persisted for you before triggering the after hook.
        }
    }

    return nil
})
```

_After the Import/Export feature I'll try to prioritize the documentation improvements._

**Reply by marcusfelix** [reply]
Cool! API rules alone is surprisingly powerful to express permissions/validations. Hooks definitely works for more complex cases. Maybe docs could improve on Server SDK (dunno if its called that way)

We may also put together something like [cloudflare/production-saas](https://github.com/cloudflare/production-saas) in the future

Thanks!!
